### ISSUE TITLE
`OverlayPortal` doesn't interleave overlay in the widget semantics

### ISSUE THREAD

**Issue by marcianx** [flutter/flutter#134456]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

This is toward making non-modal floating widgets accessible. Use-cases:
1. A widget opens a tooltip with focusable content within it. I'd like screen reader navigation to end up within the content.
2. A math widget opens a custom floating on-screen keyboard. I'd like the screen reader's virtual cursor to navigate directly from the widget onto this on-screen keyboard.

For both of these, the `SemanticsNode` tree needs to interleave the overlay right after the widget on which it's anchored.

`OverlayPortal` seems perfect for this and _does_ automatically set the focus traversal order correctly, but it seems to not order the semantics tree to match. Here's a [repro on DartPad](https://dartpad.dev/?id=6d2c070bd7321ce52c697d6085a32365) (see Console for semantics output).

**Expected semantics order:** "button1", "floating-pause", "button2"

**Observed semantics order:** "floating-pause", "button1", "button2"

### Expected results

See above.

### Actual results

See above.

### Code sample

<details><summary>Code sample</summary>

https://dartpad.dev/?id=6d2c070bd7321ce52c697d6085a32365

```dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

void main() async {
  runApp(
    MaterialApp(
      theme: ThemeData(useMaterial3: true),
      title: 'Test',
      home: const Scaffold(
        body: MyWidget(),
      ),
    ),
  );

  SemanticsBinding.instance.ensureSemantics();
  WidgetsBinding.instance.addPostFrameCallback((_) {
    debugDumpSemanticsTree();
  });
}

class MyWidget extends StatefulWidget {
  const MyWidget();

  @override
  State<MyWidget> createState() => _MyWidgetState();
}

class _MyWidgetState extends State<MyWidget> {
  final OverlayPortalController _controller = OverlayPortalController();

  @override
  Widget build(BuildContext context) {
    _controller.show();
    return FocusTraversalGroup(
      // Necessary to ensure that floating-pause is between button1 and button2
      // in the focus traversal order. Does not help semantics order one bit.
      policy: WidgetOrderTraversalPolicy(),
      child: Row(
        children: [
          OverlayPortal(
            controller: _controller,
            overlayChildBuilder: (context) => Positioned(
                left: 100,
                top: 20,
                child: button(Icons.pause, "floating-pause")),
            child: button(Icons.play_arrow, "button1"),
          ),
          button(Icons.stop, "button2"),
        ],
      ),
    );
  }
}

Widget button(IconData icon, String label, [VoidCallback? onPressed]) {
  return IconButton(
    icon: Semantics(
      label: label,
      child: Icon(icon),
    ),
    onPressed: onPressed ?? () {},
  );
}
```

</details>

### Flutter Doctor output

N/A. See DartPad.

**Comment by goderbauer** [comment]
@LongCatIsLooong can you take a look at this?

**Comment by goderbauer** [comment]
> For both of these, the SemanticsNode tree needs to interleave the overlay right after the widget on which it's anchored.

This sounds resonable to me, I wonder if it is the right thing for all use cases, though. 

@gspencergoog You were experimenting with OverlayPortal and menus. Did you run into a11y ordering problems there? Would the order proposed above work for the menu use case?

**Comment by LongCatIsLooong** [comment]
> @LongCatIsLooong We probably have to provide a custom `visitChildrenForSemantics` implementation for one of the helper OverlayPortal RenderObjects...

Yeah that sounds like a more reasonable default behavior. But what would the right order be, before the regular child or after?

**Comment by goderbauer** [comment]
The use case mentioned in the issue speculates that it should come after the regular child, which on the surface sounds reasonable. Would be interesting to look at some other potential use cases though, for example the menu one.

**Comment by gspencergoog** [comment]
In this PR: https://github.com/flutter/flutter/pull/130534 that converts menus to overlay portal, It would be nice to see if an updated semantics order would fix the test at https://github.com/gspencergoog/flutter/blob/39044d800ebd3ee7454f14f901c97369cfe28c51/packages/flutter/test/material/menu_anchor_test.dart#L3232 

Currently, (both in the main branch and this PR) [this code](https://github.com/gspencergoog/flutter/blob/39044d800ebd3ee7454f14f901c97369cfe28c51/packages/flutter/lib/src/material/menu_anchor.dart#L1936) sets the semantics node to have "expanded" set to true, but that doesn't get propagated when the semantics are merged in the PR. Without overlay portal, it is merged properly.

**Comment by marcianx** [comment]
IMO, the default behavior should match focus traversal's default behavior when traversing in widget order, which is currently after the regular/anchor child.

**Comment by LongCatIsLooong** [comment]
@marcianx in this example

> A math widget opens a custom floating on-screen keyboard. I'd like the screen reader's virtual cursor to navigate directly from the widget onto this on-screen keyboard.

If the math widget is in a `ListView` but at the same time it is outside of the `ListView`'s semantics boundary then chances are the math widget's semantic nodes will be dropped from the tree. If the on-screen keyboard is visible, what should the semantics tree look like?

**Comment by marcianx** [comment]
The idea is that you have a math field accompanied with a math keyboard toggle button. The toggle button triggers a floating math keyboard to appear at the bottom of the screen. This math keyboard needs to be interleaved between the field and the toggle button both in the focus and semantic orders to allow the user to easily navigate between it and the math field and keyboard toggle button. When the focus moves altogether out of these collections of widgets, then the floating keyboard closes automatically (which is also why the keyboard is interleaved between these two widgets):
![image](https://github.com/flutter/flutter/assets/370338/7fb4f90a-fa74-4dda-9b0a-4189d662ae0d)
